File: backend/Model/Invoice.cs
namespace YLWorks.Model
{
    public class Invoice: BaseEntity
    {
        public Guid Id { get; set; }
        public string InvoiceNo { get; set; } = string.Empty;
        public Guid? DeliveryOrderId { get; set; }
        public DeliveryOrder DeliveryOrder { get; set; }
        public Guid? ClientId { get; set; }
        public Company? Client { get; set; }
        public Guid? SupplierId { get; set; }
        public Company? Supplier { get; set; }
        public string? ProjectCode { get; set; }
        public Project? Project { get; set; }
        public Guid? PurchaseOrderId { get; set; }
        public PurchaseOrder? PurchaseOrder { get; set; }
        public Guid? QuotationId { get; set; }
        public Quotation? Quotation { get; set; }

        public DateTime InvoiceDate {  get; set; }
        public DateTime DueDate { get; set; }
       
        public string? Remarks { get; set; }
        public string Status { get; set; } = "Draft";  // Paid, Overdue, Upcoming, Cancelled, Partially Paid, Unpaid, Refunded, Draft

        public decimal? Discount { get; set; }
        public decimal Gross { get; set; }
        public decimal TotalAmount { get; set; }
        public decimal? PaidAmount { get; set; }

        public string? Notes { get; set; }
        public string? Description { get; set; }
        public string? TermsAndConditions { get; set; }
        public string? BankDetails { get; set; }

        public Guid CreatedById { get; set; }
        public User CreatedBy { get; set; }

        public ICollection<InvoiceItem> InvoiceItems { get; set; } = new List<InvoiceItem>();
        public ICollection<AttachmentDto> Attachments { get; set; } = new List<AttachmentDto>();
        public ICollection<Payments> Payments { get; set; } = new List<Payments>();


    }

    public class InvoiceItem: BaseEntity
    {
        public Guid Id { get; set; }
        public Guid InvoiceId { get; set; }
        public Invoice Invoice { get; set; }

        public string Item { get; set; }
        public string Description { get; set; } = string.Empty;
        public int Quantity { get; set; }
        public string Unit { get; set; } = "Unit";
        public decimal UnitPrice { get; set; }
        public decimal Discount { get; set; } // Percentage
        public decimal TotalAmount { get; set; }
    }

    public class InvoiceItemBase
    {
        public string Item { get; set; }
        public string Description { get; set; } = string.Empty;
        public int Quantity { get; set; }
        public string Unit { get; set; } = "Unit";
        public decimal UnitPrice { get; set; }
        public decimal Discount { get; set; }
        public decimal TotalAmount { get; set; }
    }

    // Used for Create: No ID needed as all items are new
    public class InvoiceItemRequest : InvoiceItemBase { }

    // Used for Update: Optional ID to differentiate between 'Edit' and 'Add New'
    public class UpdateInvoiceItemRequest : InvoiceItemBase
    {
        public Guid? Id { get; set; }
    }

    public class CreateInvoiceRequest
    {
        public string InvoiceNo { get; set; } = string.Empty;
        public Guid? DeliveryOrderId { get; set; }
        public Guid? ClientId { get; set; }
        public Guid? SupplierId { get; set; }
        public string? ProjectCode { get; set; }
        public Guid? PurchaseOrderId { get; set; }
        public Guid? QuotationId { get; set; }
        public DateTime InvoiceDate { get; set; }
        public DateTime DueDate { get; set; }
        public decimal Gross { get; set; }
        public decimal? Discount { get; set; }
        public decimal TotalAmount { get; set; }

        public string? Description { get; set; }
        public string? TermsConditions { get; set; }
        public string? BankDetails { get; set; }
        public string? Remarks { get; set; }
        public string? Notes { get; set; }

        public List<InvoiceItemRequest> InvoiceItems { get; set; } = new();
    }

    public class UpdateInvoiceRequest : CreateInvoiceRequest
    {
        public Guid Id { get; set; }

        // Replaces ICollection<QuotationItemRequest> with the Update version
        public new List<UpdateInvoiceItemRequest>? InvoiceItems { get; set; }
    }

    public class UpdateInvoiceStatusRequest
    {
        public Guid InvoiceId { get; set; }
        public string Status { get; set; } = string.Empty;
        // Example: Planned, InProgress, OnHold, Completed, Cancelled
    }

    public class MarkInvoicePaidRequest
    {
        public Guid InvoiceId { get; set; }
        public decimal Amount { get; set; }
        public string? PaymentMethod { get; set; }
        public string? Notes { get; set; }
        public Guid ProcessedById { get; set; } // Who processed the payment
    }

    public class InvoiceSummaryDto
    {
        public decimal TotalAmount { get; set; }
        public decimal TotalPercentage { get; set; }

        public decimal? PaidAmount { get; set; }
        public decimal PaidPercentage { get; set; }

        public decimal? PendingAmount { get; set; }
        public decimal PendingPercentage { get; set; }

        public decimal? OverdueAmount { get; set; }
        public decimal OverduePercentage { get; set; }
    }

}

File: frontend/src/app/models/Invoice.ts
import { AttachmentDto } from './AttachmentDto';
import { BaseModel } from './BaseModel';
import { CompanyDto } from './Company';
import { DeliveryOrderDto } from './DeliveryOrder';
import { PaymentDto } from './Payments';
import { ProjectDto } from './Project';
import { PurchaseOrderDto } from './PurchaseOrder';
import { QuotationDto } from './Quotation';
import { UserDto } from './User';

// Invoice model
export interface InvoiceDto extends BaseModel {
  invoiceNo: string;
  deliveryOrderId: string;
  deliveryOrder: DeliveryOrderDto;
  clientId: string;
  client: CompanyDto;
  supplierId: string;
  supplier: CompanyDto;
  projectId: string;
  project: ProjectDto;
  purchaseOrderId: string;
  purchaseOrder: PurchaseOrderDto;
  quotationId: string;
  quotation: QuotationDto;
  invoiceDate: Date;
  dueDate: Date;
  terms: string;
  remarks: string;
  status: string;
  type: string;
  discount: number;
  gross: number;
  totalAmount: number;
  paidAmount: number;
  notes: string;
  termsAndConditions: string;
  bankDetails: string;
  attachment: string;
  createdById: string;
  createdBy: UserDto;
  invoiceItems: InvoiceItem[];
  payments: PaymentDto[];
}

// Invoice Item model
export interface InvoiceItem extends BaseModel {
  invoiceId: string;
  invoice: InvoiceDto;
  item: string;
  description: string;
  quantity: number;
  unit: string;
  unitPrice: number;
  discount: number;
  amount: number;
}

// Create Invoice Request
export interface CreateInvoiceRequest {
  invoiceNo: string;
  deliveryOrderId: string;
  clientId: string;
  supplierId: string;
  projectId: string;
  purchaseOrderId: string;
  quotationId: string;
  type: string;
  invoiceDate: Date;
  dueDate: Date;
  gross: number;
  discount: number;
  totalAmount: number;
  terms: string;
  termsAndConditions: string;
  bankDetails: string;
  remarks: string;
  notes: string;
  attachment: File;
  invoiceItems: InvoiceItemRequest[];
}

export interface InvoiceItemRequest {
  item: string;
  description: string;
  quantity: number;
  unit: string;
  unitPrice: number;
  discount: number;
  amount: number;
}

export interface UpdateInvoiceRequest {
  id: string;
  invoiceNo: string;
  deliveryOrderId: string;
  clientId: string;
  supplierId: string;
  projectId: string;
  purchaseOrderId: string;
  quotationId: string;
  invoiceDate: Date;
  dueDate: Date;
  gross: number;
  discount: number;
  totalAmount: number;
  terms: string;
  termsAndConditions: string;
  bankDetails: string;
  remarks: string;
  notes: string;
  attachment: File;
  invoiceItems: InvoiceItemUpdateRequest[];
}

export interface InvoiceItemUpdateRequest extends InvoiceItemRequest {
  id: string;
}

export interface UpdateInvoiceStatusRequest {
  invoiceId: string;
  status: string;
}

export interface MarkInvoicePaidRequest {
  invoiceId: string;
  amount: number;
  paymentMethod: string;
  notes: string;
  processedById: string;
}

export interface InvoiceSummaryDto {
  totalAmount: number;
  totalPercentage: number;
  paidAmount: number | null;
  paidPercentage: number;
  pendingAmount: number | null;
  pendingPercentage: number;
  overdueAmount: number | null;
  overduePercentage: number;
}
